File: quicksort.cpp
#include <iostream>

//void printArray(int array[], int sizeOfArray) {
//    for (int i = 0; i < sizeOfArray; i++) {
//        std::cout << array[i] << " ";
//    }
//    std::cout << std::endl;
//}

int partition(int array[], int left, int right, int sizeOfArray) {
    int pivot = array[left];
    int l_p = left - 1;
    int r_p = right + 1;
    while (true) {
        while (true) {
            l_p++;
            if (array[l_p] >= pivot)
                break;
        }
        while (true) {
            r_p--;
            if (array[r_p] <= pivot)
                break;
        }
        if (l_p < r_p) {
            int reml_p = array[l_p];
            array[l_p] = array[r_p];
            array[r_p] = reml_p;
        }
        else {
            return r_p;
        }
    }
}

void quickSort(int array[], int left, int right, int sizeOfArray) {
    if (left < right) {
        int q = partition(array, left, right, sizeOfArray);
        quickSort(array, left, q, sizeOfArray);
        quickSort(array, q + 1, right, sizeOfArray);
    }
}

int main() {
    //int array[] = {85,52,23,15,34,69,16,77,59,67,41,45,37,51,96,68,92,61,55,27,53,85,64,83,16,47,72,4,51,69,59,21,49,65,86,15,30,67,62,29,20,53,84,8,59,48,19,71,32,63};
    int array[] = { 85,52,23,79,34,69,16,12,14,15,18,19 };
    int sizeArr = sizeof(array) / sizeof(array[0]);
    quickSort(array, 0, sizeArr - 1, sizeArr);
}

File: sorts/quicksort/quicksort.py
arr = [ 5, 3, 10, 2, 4, 80, 21, 54, 24, 1, 58, 92, 55, 54, 49, 42, 59, 99, 813, 54, 58, 9, 87, 9, 53, 12, 82, 45, 69, 66, 82, 8, 67, 21, 25, 3, 18, 42, 42, 11, 61, 61, 70, 40, 49, 37, 82, 83, 22, 87, 18, 65, 48, 33, 54, 25, 71, 14, 3, 80, 57, 50, 48, 1, 21, 38, 86, 63, 37, 43, 78, 62, 67, 40, 88, 35, 1, 96, 24, 73, 64, 86, 21, 83, 83, 57, 41, 16, 81, 38, 88, 65, 44, 60, 63, 7, 47, 53, 58, 98, 56, 78]

def partition(arr, l, r):
    pivot = arr[l]
    l_p = l - 1
    r_p = r + 1
    while True:
        while True:
            l_p = l_p + 1
            if arr[l_p] >= pivot : break
        while True:
            r_p = r_p - 1
            if arr[r_p] <= pivot : break
        if (l_p < r_p):
            arr[l_p], arr[r_p] = arr[r_p], arr[l_p]
        else:
            return r_p

def quick_sort(arr, l, r):
    if (l < r):
        q = partition(arr, l, r)
        quick_sort (arr, l, q)
        quick_sort (arr, q+1, r)

quick_sort(arr, 0, len(arr) - 1)

print(arr)
def check_if_sorted(arr):
    for n in range(len(arr) - 1):
        if arr[n] > arr[n + 1]:
            return False
    return True
print(check_if_sorted(arr))
